### PR TITLE
Improving coverage for croniter

### PR DESCRIPTION
**Pull request by sg3-141-592**
`fuzz_iter.py`
- Added coverage of optional `hash_id` parameter
- Added call to `is_valid` method
- Added call to `get_next` & `get_prev` methods for the returned croniter iterator

`fuzz_match.py`
- Added fuzzing of `croniter.match`

`fuzz_range.py`
- Added fuzzing of `croniter.range`

### CODE REVIEW

**Comment by jonathanmetzman** [comment]
@DavidKorczynski do you approve?

File: projects/croniter/fuzz_iter.py [hunk 1]
```diff
@@ -1,5 +1,5 @@
 #!/usr/bin/python3
-# Copyright 2022 Google LLC
+# Copyright 2023 Google LLC
```
Comment (DavidKorczynski) [inline projects/croniter/fuzz_iter.py:2]: No need to bump the year -- we keep the copyright from when it was introduced.

File: projects/croniter/fuzz_iter.py [hunk 2]
```diff
@@ -17,32 +17,32 @@
 import sys
 
 from datetime import datetime
-import croniter
 
+with atheris.instrument_imports():
```
Comment (DavidKorczynski) [inline projects/croniter/fuzz_iter.py:21]: Any particular reason changing this? I personally prefer `instrument_all`.

**Review by DavidKorczynski — CHANGES_REQUESTED** [review]
Quick one for now -- I'll take a bit deeper look shortly

File: projects/croniter/fuzz_iter.py (hunk 2 above)
Comment (sg3-141-592) [reply projects/croniter/fuzz_iter.py:21]: This was a misguided attempt by me to reduce the coverage surface area a bit. Croniter is a fairly small library and is pulling in a lot of code coverage from its dependencies `dateutil` and `six`.

image

Comment (sg3-141-592) [reply projects/croniter/fuzz_iter.py:21]: This was a misguided attempt by me to reduce the coverage surface area a bit. Croniter is a fairly small library and is pulling in a lot of code coverage from its dependencies `dateutil` and `six`.

image

File: projects/croniter/fuzz_iter.py (hunk 1 above)
Comment (sg3-141-592) [reply projects/croniter/fuzz_iter.py:2]: Copyright year reverted back to `2022`

File: projects/croniter/fuzz_iter.py [hunk 3]
```diff
@@ -17,32 +17,32 @@
...
+    itr.get_prev(base)
   except (croniter.CroniterBadCronError, croniter.CroniterBadDateError) as e:
     pass
   except NameError as e:
     # Catch https://github.com/kiorky/croniter/blob/bb5a45196e5f8f15fd0890f4ee5e9697671a3fe2/src/croniter/croniter.py#L781
     if not "'exc' is not defined" in str(e):
       raise e
 
-
-def main():
```
Comment (DavidKorczynski) [inline projects/croniter/fuzz_iter.py:42]: I'd probably prefer to keep main functions -- any particular reason to change the style?

Comment (sg3-141-592) [reply projects/croniter/fuzz_iter.py]: I don't, I've tidied up the `fuzz_iter.py` to fix all the unneeded changes (whitespace, def main removed etc).

**Review by DavidKorczynski — APPROVED** [review]

File: projects/croniter/fuzz_match.py [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+def TestOneInput(data):
+  fdp = atheris.FuzzedDataProvider(data)
+  cron_str = fdp.ConsumeString(50)
+  testdate = RandomDateTime(fdp)
+  try:
+    croniter.croniter.match(cron_str, testdate)
+  except croniter.CroniterBadCronError as e:
+    pass
+
+if __name__ == "__main__":
```
Comment (DavidKorczynski) [inline projects/croniter/fuzz_match.py:36]: Apologies for early approval, just noticed this -- for consistency, let's please keep the same style using `main`.

**Review by DavidKorczynski — CHANGES_REQUESTED** [review]

File: projects/croniter/fuzz_match.py (hunk 4 above)
Comment (sg3-141-592) [reply projects/croniter/fuzz_match.py]: Updated to the `def main():` style

File: projects/croniter/fuzz_match.py [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+  testdate = RandomDateTime(fdp)
+  try:
+    croniter.croniter.match(cron_str, testdate)
+  except croniter.CroniterBadCronError as e:
+    pass
+
+
+def main():
+  atheris.instrument_all()
+  atheris.Setup(sys.argv, TestOneInput, enable_python_coverage=True)
```
Comment (DavidKorczynski) [inline projects/croniter/fuzz_match.py:38]: no need for `enable_python_coverage=True`

File: projects/croniter/fuzz_match.py [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+import atheris
+import sys
+
+import datetime
+import croniter
+
+def RandomDateTime(fdp):
+    return datetime.datetime.now() + fdp.ConsumeProbability() * datetime.timedelta(days=200000)
+
+@atheris.instrument_func
```
Comment (DavidKorczynski) [inline projects/croniter/fuzz_match.py:25]: please remove

File: projects/croniter/fuzz_range.py [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+      idx += 1
+      if idx > 10:
+        break
+  except croniter.CroniterBadCronError as e:
+    pass
+
+
+def main():
+  atheris.instrument_all()
+  atheris.Setup(sys.argv, TestOneInput, enable_python_coverage=True)
```
Comment (DavidKorczynski) [inline projects/croniter/fuzz_range.py:43]: no need for `enable_python_coverage`

File: projects/croniter/fuzz_range.py [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+import atheris
+import sys
+
+import datetime
+import croniter
+
+def RandomDateTime(fdp):
+    return datetime.datetime.now() + fdp.ConsumeProbability() * datetime.timedelta(days=200000)
+
+@atheris.instrument_func
```
Comment (DavidKorczynski) [inline projects/croniter/fuzz_range.py:25]: please remove since we use `instrument_all`

**Review by DavidKorczynski — CHANGES_REQUESTED** [review]

File: projects/croniter/fuzz_match.py (hunk 5 above)
Comment (sg3-141-592) [reply projects/croniter/fuzz_match.py]: Removed

File: projects/croniter/fuzz_match.py (hunk 6 above)
Comment (sg3-141-592) [reply projects/croniter/fuzz_match.py]: Removed

File: projects/croniter/fuzz_range.py (hunk 8 above)
Comment (sg3-141-592) [reply projects/croniter/fuzz_range.py]: Removed

File: projects/croniter/fuzz_range.py (hunk 7 above)
Comment (sg3-141-592) [reply projects/croniter/fuzz_range.py]: Removed

**Review by DavidKorczynski — APPROVED** [review]

**Pull request merged**